src/coreclr/vm/unsafeaccessors.cpp

```diff
@@ -75,16 +77,220 @@ namespace
+    }
+
+    void ProcessUnsafeAccessorTypeAttributes(GenerationContext& cxt)
+    {
+        STANDARD_VM_CONTRACT;
+
+        // Acquire attribute name to search for.
+        const char* typeAttrName = GetWellKnownAttributeName(WellKnownAttribute::UnsafeAccessorTypeAttribute);
+
+        uint32_t attrCount = 0;
+
+        // Allocate memory to store the updated types.
+        NewArrayHolder<MethodTable*> heapAlloc;
+        MethodTable* stackAlloc[12]; // Use the stack alloc the majority of the time.
+        MethodTable** translations = stackAlloc;
+
+        // Determine the max parameter count. +1 for the return value, which is always index zero.
+        const uint32_t totalParamCount = cxt.DeclarationMetaSig.NumFixedArgs() + 1;
+
+        // Inspect all parameters to the declaration method for UnsafeAccessorTypeAttribute.
+        IMDInternalImport *pInternalImport = cxt.Declaration->GetModule()->GetMDImport();
+        HENUMInternalHolder hEnumParams(pInternalImport);
+        hEnumParams.EnumInit(mdtParamDef, cxt.Declaration->GetMemberDef());
+        mdParamDef currParamDef = mdParamDefNil;
+        while (hEnumParams.EnumNext(&currParamDef))
+        {
+            const void *pData;
+            ULONG cbData;
+            HRESULT hr = IfFailThrow(pInternalImport->GetCustomAttributeByName(currParamDef, typeAttrName, &pData, &cbData));
+            if (hr != S_OK)
+                continue;
+
+            // The first time we find an attribute, we need to initialize
+            // the translations array and might need to allocate a larger array
+            // to store the translations.
+            if (attrCount == 0)
+            {
+                if (totalParamCount > ARRAY_SIZE(stackAlloc))
+                {
+                    heapAlloc = new MethodTable*[totalParamCount];
+                    translations = heapAlloc;
+                }
+                memset(translations, 0, sizeof(translations[0]) * totalParamCount);
+            }
+
+            // Parse the attribute data.
+            CustomAttributeParser cap(pData, cbData);
+            IfFailThrow(cap.ValidateProlog());
+            LPCUTF8 typeString;
+            ULONG typeStringLen;
+            IfFailThrow(cap.GetNonNullString(&typeString, &typeStringLen));
+
+            StackSString typeStringUtf8{ SString::Utf8, typeString, typeStringLen };
+
+            // Pass the string in the attribute to similar logic as Type.GetType(String).
+            // The below API creates a dependency between the returned type and the requesting
+            // assembly for the purposes of lifetime tracking of collectible types.
+            TypeHandle typeHandle = TypeName::GetTypeReferencedByCustomAttribute(
+                typeStringUtf8.GetUnicode(),
+                cxt.Declaration->GetAssembly(),
+                cxt.Declaration /* unsafeAccessorMethod */);
+            _ASSERTE(!typeHandle.IsNull());
+
+            MethodTable* targetType = typeHandle.AsMethodTable();
```

Comment:
The type handle returned by `GetTypeReferencedByCustomAttribute` can be TypeDesc. For example, consider this:
```
using System.Runtime.CompilerServices;

class Test
{
    static void Main()
    {
        object o = null;
        M2(null, ref o);
    }

    [UnsafeAccessor(UnsafeAccessorKind.Method, Name = "M")]
    static extern void M2(Test dummy, [UnsafeAccessorType("Test&")] ref object t);

    static void M(ref Test t) { }
}
```

Reply:
(This fails with `Assert failure(PID 33220 [0x000081c4], Thread: 40356 [0x9da4]): !IsTypeDesc()`.)